coap/adapter.go

```diff
@@ -1,61 +1,85 @@
-package adapter
+package coap
 
 import (
-	"encoding/json"
-	"log"
-	"net"
+	"errors"
+	"sync"
 
-	"github.com/dustin/go-coap"
 	"github.com/mainflux/mainflux"
+	"github.com/mainflux/mainflux/coap/nats"
 	broker "github.com/nats-io/go-nats"
-	"go.uber.org/zap"
 )
 
-const protocol string = "coap"
+const (
+	key       string = "key"
+	channel   string = "id"
+	protocol  string = "coap"
+	maxPktLen int    = 1500
+	network          = "udp"
+)
 
-type Observer struct {
-	conn    *net.UDPConn
-	addr    *net.UDPAddr
-	message *coap.Message
-}
+var (
+	errBadRequest = errors.New("bad request")
+	errBadOption  = errors.New("bad option")
+)
 
-type CoAPAdapter struct {
-	obsMap map[string][]Observer
-	logger *zap.Logger
-	pub    mainflux.MessagePublisher
+// AdapterService struct represents CoAP adapter service implementation.
+type adapterService struct {
+	pubsub nats.Service
+	Subs   map[string]nats.Observer
+	mu     sync.Mutex
 }
 
-// NewCoAPAdapter creates new CoAP adapter struct
-func NewCoAPAdapter(logger *zap.Logger, pub mainflux.MessagePublisher) *CoAPAdapter {
-	ca := &CoAPAdapter{
-		logger: logger,
-		pub:    pub,
-		obsMap: make(map[string][]Observer),
+// New creates new CoAP adapter service struct.
+func New(pubsub nats.Service) Service {
+	return &adapterService{
+		pubsub: pubsub,
+		Subs:   make(map[string]nats.Observer),
+		mu:     sync.Mutex{},
 	}
-
-	return ca
 }
 
-// Serve function starts CoAP server
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
+func (svc *adapterService) Publish(msg mainflux.RawMessage) error {
+	if err := svc.pubsub.Publish(msg); err != nil {
+		switch err {
+		case broker.ErrConnectionClosed, broker.ErrInvalidConnection:
+			return ErrFailedConnection
+		default:
+			return ErrFailedMessagePublish
+		}
+	}
+	return nil
 }
 
-// BridgeHandler functions is a handler for messages received via NATS
-func (ca *CoAPAdapter) BridgeHandler(nm *broker.Msg) {
-	log.Printf("Received a message: %s\n", string(nm.Data))
-
-	// And write it into the database
-	m := mainflux.RawMessage{}
-	if len(nm.Data) > 0 {
-		if err := json.Unmarshal(nm.Data, &m); err != nil {
-			log.Println("Can not decode adapter msg")
-			return
-		}
+func (svc *adapterService) Subscribe(chanID, clientID string, ch chan mainflux.RawMessage) error {
+	sub, err := svc.pubsub.Subscribe(chanID, ch)
+	if err != nil {
+		return ErrFailedSubscription
+	}
+	// Remove if entry already exists.
+	if _, ok := svc.Subs[clientID]; ok == true {
+		svc.Unsubscribe(clientID)
+	}
+	svc.mu.Lock()
+	svc.Subs[clientID] = nats.Observer{
```

Comment:
Maybe you should encapsulate set and delete methods for this map that will be thread safe (use lock).